sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -291,13 +359,194 @@ class KernelProgramCache {
 
   template <typename KeyT, typename ValT>
   void saveKernel(KeyT &&CacheKey, ValT &&CacheVal) {
+
+    if (SYCLConfig<SYCL_IN_MEM_CACHE_EVICTION_THRESHOLD>::
+            isProgramCacheEvictionEnabled()) {
+
+      ur_program_handle_t Program = std::get<3>(CacheVal);
+      // Save kernel in fast cache only if the corresponding program is also
+      // in the cache.
+      {
+        auto LockedCache = acquireCachedPrograms();
+        auto &ProgCache = LockedCache.get();
+        if (ProgCache.ProgramSizeMap.find(Program) ==
+            ProgCache.ProgramSizeMap.end())
+          return;
+      }
+
+      // Save reference between the program and the fast cache key.
+      MProgramToKernelFastCacheKeyMap[Program].emplace_back(CacheKey);
+    }
+
     std::unique_lock<std::mutex> Lock(MKernelFastCacheMutex);
     // if no insertion took place, thus some other thread has already inserted
     // smth in the cache
     traceKernel("Kernel inserted.", std::get<3>(CacheKey), true);
     MKernelFastCache.emplace(CacheKey, CacheVal);
   }
 
+  // Evict programs from cache to free up space.
+  void evictPrograms(size_t DesiredCacheSize, int CurrentCacheSize) {
+
+    // Figure out how many programs from the beginning we need to evict.
+    // [FIXME] Will the race on MCachedPrograms.Cache.empty() be a problem?
+    size_t EvictSize = CurrentCacheSize - DesiredCacheSize;
+    if (EvictSize <= 0 || MCachedPrograms.Cache.empty())
+      return;
+
+    // Evict programs from the beginning of the cache.
+    {
+      std::lock_guard<std::mutex> Lock(MProgramEvictionListMutex);
+
+      // Traverse the eviction list and remove the LRU programs.
+      // The LRU programs will be at the front of the list.
+      while (EvictSize > DesiredCacheSize && !MEvictionList.empty()) {
+        ProgramCacheKeyT CacheKey = MEvictionList.MProgramEvictionList.front();
+        auto LockedCache = acquireCachedPrograms();
+        auto &ProgCache = LockedCache.get();
+        auto It = ProgCache.Cache.find(CacheKey);
+        if (It != ProgCache.Cache.end()) {
+          // We are about to remove this program now.
+          // (1) Remove it from KernelPerProgram cache.
+          // (2) Remove corresponding entries from KernelFastCache.
+          // (3) Remove it from ProgramCache KeyMap.
+          // (4) Remove it from the ProgramCache.
+          // (5) Remove it from ProgramSizeMap.
+          // (6) Update the cache size.
+
+          // Remove entry from the KernelsPerProgram cache.
+          ur_program_handle_t NativePrg = It->second->Val;
+          {
+            auto LockedCacheKP = acquireKernelsPerProgramCache();
+            // List kernels that are to be removed from the cache, if tracing is
+            // enabled.
+            if (SYCLConfig<SYCL_CACHE_TRACE>::isTraceInMemCache()) {
+              for (const auto &Kernel : LockedCacheKP.get()[NativePrg])
+                traceKernel("Kernel evicted.", Kernel.first);
```

Comment:
Is it because of the performance? If so, the I/O will happen only during debugging, if we set `SYCL_CACHE_TRACE`. There shouldn't be any performance hit otherwise.

Reply:
Mainly performance, yes. But also, if for some reason I/O blocks, our lock will also be locked until I/O unblocks. I get that it's only during debugging, so I'm not strongly against it, but if there's an easy alternative, it would be nice.